Fix and extend Travis builds

daviddrysdale opened the pull request:
I've been looking at the Travis CI setup for libgd, and I had some suggestions:

 - The libgd `configure` script looks to have a bug where it turns on `with_webp` even when its check fails (and this check fails on Travis's systems, causing failures for `make check`).
 - With WebP detection fixed, the `make check` target works OK and so can be re-enabled as something that should fail a build.
 - The codebase is [ASan](http://clang.llvm.org/docs/AddressSanitizer.html) and (mostly) [UBSan](http://clang.llvm.org/docs/UsersManual.html#controlling-code-generation) clean, so it would be nice to add build targets to keep it that way.  I also added a code coverage generation [build](https://coveralls.io/builds/4045988), but that needs the project owners to click a couple of things at [Coveralls](https://coveralls.io/).

(There's a [set of builds](https://travis-ci.org/daviddrysdale/libgd/builds/89238693) visible at Travis with my changes.)

daviddrysdale commented:
I've rejigged the configure test along the lines you suggested -- thanks.  (I've also commented out the Coveralls build, as it gives a failed build until the libgd project is set up there.)

vapier commented:
i've rewritten the configure script in pull request #193, so you probably want to drop the webp commit from this series

daviddrysdale commented:
I've updated the branch to be based on pull request #193, and added a couple more sanitizers.  Those exposed a few leaks, mostly in the test code, which I've added fixes for.

vapier commented:
i actually merged the configure updates yesterday ;)

vapier commented:
i've pushed your "Fail Travis builds if make check fails" to master now by hand

vapier commented on .travis.yml:
```diff
@@ -1,15 +1,51 @@
-os:
-  - linux
-  - osx
+matrix:
```
this matrix is way over the top imo.  there's no need to split up the sanitizer logic this much.  just do a normal build and a debug build, and for the debug build, include all the sanitizers at once.

vapier commented:
i cherry picked the tiff leak fix, and fixed another one at the same time in the same func -- see commit 4e61c9b0212b4a8a46ce4848252ab2926c7c9d06

vapier commented:
add i've cherry picked your test leak fixes -- see commit 05d70f6a072e480186414af34475ed43d71f035c

vapier commented:
this needs a pretty big rebase.  feel free to re-open if you get around to it.

vapier commented:
for posterity, this has been refreshed via #235
